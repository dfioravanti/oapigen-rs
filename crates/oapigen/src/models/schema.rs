use proc_macro2::TokenStream;
use quote::{quote, ToTokens};
use std::hash::Hash;
use std::{collections::HashMap, fmt};

/// Collects all the imports that were already added to the code.
/// It is used to determine if an import should be added or not.
/// For example assume that two different structs need
/// chrono::DateTime then we add "chrono::DateTime" to the Imports
/// for the first struct that we hit and we will not add it in the second case since we already have it.
pub type Imports = HashMap<String, TokenStream>;

#[derive(Debug)]
pub enum CurrentType {
    Type,
    Const,
    Enum,
    Vec,
    Struct,
}

/// SchemaAsRust represents a rust type that was inferred from a OpenAPI specification.
/// It contains all the information needed to render the type correctly.
#[derive(Debug)]
pub struct SchemaAsRust {
    /// The name of the field as a TokenStream
    pub tokenized_name: TokenStream,
    /// The type represented as a TokenStream
    pub tokenized_type: TokenStream,
    /// The imports needed to make the type compile.
    pub imports: Imports,

    pub current_type: CurrentType,
}
impl Eq for SchemaAsRust {}
impl PartialEq for SchemaAsRust {
    fn eq(&self, other: &Self) -> bool {
        self.to_token_stream().to_string() == other.to_token_stream().to_string()
    }
}

impl Hash for SchemaAsRust {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.to_token_stream().to_string().hash(state);
    }
}

impl fmt::Display for SchemaAsRust {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let imports = self
            .imports
            .iter()
            .fold(TokenStream::new(), |mut acc, (_, v)| {
                acc.extend(v.clone());
                acc
            });

        if imports.is_empty() {
            write!(f, "{}", self.to_token_stream())
        } else {
            write!(f, "{} \n {}", imports, self.to_token_stream())
        }
    }
}

impl ToTokens for SchemaAsRust {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let generated_tokens = match self.current_type {
            CurrentType::Type => tokenize_type(self),
            CurrentType::Const => todo!(),
            CurrentType::Enum => todo!(),
            CurrentType::Vec => todo!(),
            CurrentType::Struct => todo!(),
        };

        tokens.clone_from(&generated_tokens);
    }
}

fn tokenize_type(tokenized_schema: &SchemaAsRust) -> TokenStream {
    let tokenized_name = &tokenized_schema.tokenized_name;
    let tokenized_type = &tokenized_schema.tokenized_type;

    quote! { struct #tokenized_name(#tokenized_type); }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    #[test]
    fn test_parsed_schema_display() {
        let tokenized_schema = quote! {
            DateTime
        };

        let mut imports = HashMap::new();
        imports.insert(
            "chrono::DateTime".to_string(),
            quote! { use chrono::DateTime; },
        );

        let parsed_schema = SchemaAsRust {
            tokenized_name: quote! { time },
            tokenized_type: tokenized_schema,
            imports,
            current_type: CurrentType::Type,
        };

        let formatted_parsed_schema = format!("{}", parsed_schema);

        assert_yaml_snapshot!(formatted_parsed_schema);
    }
}
