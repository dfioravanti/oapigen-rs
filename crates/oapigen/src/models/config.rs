use std::path::PathBuf;

pub enum DateTimeLibraries {
    Chrono,
    Jiff,
}

pub struct Libraries {
    pub datetime: DateTimeLibraries,
}

impl Default for Libraries {
    fn default() -> Self {
        Libraries {
            datetime: DateTimeLibraries::Chrono,
        }
    }
}

#[derive(Default)]
pub struct Config {
    pub output_path: PathBuf,
    pub libraries: Libraries,
}
