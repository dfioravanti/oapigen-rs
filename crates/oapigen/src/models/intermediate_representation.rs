#[derive(Debug, serde::Serialize)]
pub struct Struct {
    pub name: String,
    pub description: Option<String>,
    pub fields: Vec<Field>,
}

#[derive(Debug, serde::Serialize)]
pub struct Field {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub is_required: bool,
}

pub struct ConstantField {
    pub name: String,
    pub description: Option<String>,
    pub rust_type: String,
    pub value: serde_json::Value,
}
