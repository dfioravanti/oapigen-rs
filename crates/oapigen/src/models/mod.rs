pub mod intermediate_representation;
pub mod schema;

pub use intermediate_representation::*;
pub use schema::*;
