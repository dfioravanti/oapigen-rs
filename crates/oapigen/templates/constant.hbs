#[derive(Debug)]
struct {{struct_name}};

impl Serialize for {{struct_name}} {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {   
        {{#if string_valid_value}}
        serializer.serialize_str({{string_valid_value}})
        {{/if}}
    }
}

impl<'de> Deserialize<'de> for ApiVersionV2 {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let incoming = String::deserialize(deserializer)?;

        // The core validation logic
        if incoming == {{string_valid_value}} {
            Ok(ApiVersionV2)
        } else {
            Err(de::Error::custom(format!(
                "expected string \"v2\", found \"{}\"",
                incoming
            )))
        }
    }
}
