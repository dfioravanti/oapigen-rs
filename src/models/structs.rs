use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug)]
pub struct Struct {
    pub struct_name: String,
    pub fields: Vec<Field>,
}
#[derive(Serialize, Deserialize, Debug)]
pub struct Field {
    pub field_name: String,
    pub field_type: String,
    pub comment: Option<String>,
    pub annotations: Vec<String>,
}
