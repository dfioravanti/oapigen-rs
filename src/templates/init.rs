use anyhow::{Context, Result};
use handlebars::Handlebars;
use rust_embed::Embed;

#[derive(Embed)]
#[folder = "templates"]
struct HandlebarsTemplate;

pub const STRUCT_TEMPLATE_NAME: &str = "struct";

pub fn handlebars<'a>() -> Result<Handlebars<'a>> {
    let mut handlebars = Handlebars::new();

    for template_filename in HandlebarsTemplate::iter() {
        let file = HandlebarsTemplate::get(template_filename.as_ref())
            .with_context(|| format!("Failed to load file {:?}", template_filename))?;

        let template_content = std::str::from_utf8(file.data.as_ref())
            .with_context(|| format!("Failed to load content of file {:?}", template_filename))?;

        // the template_filename includes the extension, which we need to drop
        let template_name = template_filename.clone().replace(".hbs", "");
        handlebars
            .register_template_string(template_name.as_ref(), template_content)
            .with_context(|| {
                format!(
                    "Failed to register template for file {:?}",
                    template_filename
                )
            })?;
    }

    Ok(handlebars)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_init() {
        let handlebars = handlebars();

        handlebars.unwrap();
    }
}
