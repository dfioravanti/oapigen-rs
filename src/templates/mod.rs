pub mod structs;
mod init;