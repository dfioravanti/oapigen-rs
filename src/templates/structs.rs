use crate::models;
use crate::templates::init;
use anyhow::{Context, Result};
use handlebars::Handlebars;

fn parse_struct(struct_to_be_parsed: models::Struct, handlebars: Handlebars) -> Result<String> {
    let result = handlebars
        .render(init::STRUCT_TEMPLATE_NAME, &struct_to_be_parsed)
        .with_context(|| {
            format!(
                "failed to render struct {:?}",
                struct_to_be_parsed.struct_name
            )
        })?;

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta;
    #[test]
    fn test_struct_is_rendered_correctly() {
        let handlebars = init::handlebars().unwrap();

        let input = models::Struct {
            struct_name: "User".to_string(),
            fields: vec![
                models::Field {
                    field_name: "name".to_string(),
                    field_type: "String".to_string(),
                    comment: None,
                    annotations: vec![],
                },
                models::Field {
                    field_name: "lastname".to_string(),
                    field_type: "String".to_string(),
                    comment: None,
                    annotations: vec![],
                },
            ],
        };

        let got = parse_struct(input, handlebars).unwrap();
        insta::assert_yaml_snapshot!(got);
    }
}
