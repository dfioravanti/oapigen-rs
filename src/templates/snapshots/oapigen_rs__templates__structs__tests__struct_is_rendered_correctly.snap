---
source: src/templates/structs.rs
expression: got
---
"struct User {\n            name: String,\n            lastname: String,\n}"
